src/pytroll_watchers/selector.py

```diff
@@ -0,0 +1,214 @@
+what to pass to it.
+
+
+"""
+
+import argparse
+import logging
+import time
+from contextlib import closing, contextmanager
+from functools import cache
+from pathlib import Path
+from subprocess import Popen
+
+import redis
+import yaml
+from posttroll.publisher import create_publisher_from_dict_config
+from posttroll.subscriber import create_subscriber_from_dict_config
+
+from pytroll_watchers.main_interface import configure_logging
+
+logger = logging.getLogger(__name__)
+
+@cache
+def _connect_to_redis(**kwargs):
+    return redis.Redis(**kwargs)
+
+
+class TTLDict:
+    """A simple dictionary-like object that discards items older than a time-to-live.
+
+    Not thread-safe.
+
+    Args:
+        ttl: the time to live of the stored items in integer seconds or as a timedelta instance. Cannot be less
+            than 1 second.
+        redis_params: the keyword arguments to pass to the underlying :py:class:`~redis.Redis` instance.
+
+    """
+
+    def __init__(self, ttl=300, **redis_params):
+        """Set up the instance.
+
+        Args:
+            ttl: the time to live of the stored items in integer seconds or as a timedelta instance. Cannot be less
+              than 1 second.
+            redis_params: the keyword arguments to pass to the underlying :py:class:`~redis.Redis` instance.
+        """
+        self._redis = _connect_to_redis(**redis_params)
+        self._ttl = ttl
+
+    def __getitem__(self, key):
+        """Get the value corresponding to *key*."""
+        return self._redis[key]
+
+    def __setitem__(self, key, value):
+        """Set the *value* corresponding to *key*."""
+        res = self._redis.get(key)
+        if not res:
+            self._redis.set(key, value, ex=self._ttl)
+
+
+def running_selector(selector_config, subscriber_config):
+    """Generate selected messages.
+
+    The aim of this generator is to skip messages that are duplicates to already processed messages.
+    Duplicate in this context means messages referring to the same file (even if stored in different locations).
+
+    Args:
+        selector_config: a dictionary of arguments to pass to the underlying redis instance, see
+          :py:class:`~redis.Redis`. You can also provide a ttl as an int
+          (seconds) or timedelta instance, otherwise it defaults to 300 seconds (5 minutes).
+        subscriber_config: a dictionary of arguments to pass to
+          :py:func:`~posttroll.subscriber.create_subscriber_from_dict_config`.
+
+    Yields:
+        JSON representations of posttroll messages.
+    """
+    subscriber = create_subscriber_from_dict_config(subscriber_config)
+
+    with closing(subscriber):
+        sel = TTLDict(**selector_config)
+        for msg in subscriber.recv():
+            key = msg.data["uid"]
+            try:
+                _ = sel[key]
+                logger.info(f"Discarded {str(msg)}")
```

Comment:
```suggestion
                logger.debug(f"Discarded {str(msg)}")
```
I think INFO is too high log level for discarding duplicates.

Reply:
Fixed